### PR TITLE
[cluster management] enable participant to register to multiple state models

### PR DESCRIPTION
**Pull request by kangnanli**
Currently, Participant can only register to 1 state model (e.g. MasterSlave or OnlineOffline). Change the logic to support register to multiple state models. 

Example: pass the "stateModelType" as "MasterSlave;Task", the cluster will register to "MasterSlave" and "Task".  This prepare Rocksplicator support both db management (by MasterSlave state model) and Backup/Restore (by Task state model). 
Other usage, register to "MasterSlave;OnlineOffline" is also allowed, but not recommended since better practice to separate realtime mode (ie. MasterSlave) from batch mode (ie OnlineOffline).

### CODE REVIEW

**Comment by buildinglogic** [comment]
> Can we just add a new "else if" section for "MasterSlave;Task"? This can avoid misuse and minimize changes to other models.

Two new combinations will be used for now: 
- MasterSlave;Task.   // to support db management and backup/restore
- Task.      // support export (dedup). This cluster will only use "Task" framework from helix. 

I can change to 2 new elseif to support these?

**Comment by newpoo** [comment]
Sounds good to me to have 2 new elseif.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java [hunk 1]
```diff
@@ -204,12 +215,43 @@ private Participant(String zkConnectString, String clusterName, String instanceN
...
+          port, zkConnectString, clusterName, useS3Backup, s3BucketName);
+
+      Map<String, TaskFactory> taskFactoryRegistry = new HashMap<>();
+      // TODO: register backup/restore factories
+      // taskFactoryRegistry.put("Backup", new BackupTaskFactory());
+      // taskFactoryRegistry.put("Restore", new RestoreTaskFactory());
+
+      TaskStateModelFactory helixTaskFactory =
+          new TaskStateModelFactory(helixManager, taskFactoryRegistry);
+      stateMach.registerStateModelFactory("Task", helixTaskFactory);
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java]: Can we just create and fill taskFactoryRegistry here and move registerStateModelFactory "Task" to be after? This will help remove code duplication.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java [hunk 2]
```diff
@@ -187,13 +191,20 @@ public static void disconnectHelixManager() throws Exception {
...
-                     String stateModelType, int port, String postUrl,
-                     boolean useS3Backup, String s3BucketName, boolean runSpectator) throws Exception {
+  public Participant(String zkConnectString, String clusterName, String instanceName,
+                      String stateModelType, int port, String postUrl,
+                      boolean useS3Backup, String s3BucketName, boolean runSpectator)
+      throws Exception {
     helixManager = HelixManagerFactory.getZKHelixManager(clusterName, instanceName,
         InstanceType.PARTICIPANT, zkConnectString);
 
+    StateMachineEngine stateMach = helixManager.getStateMachineEngine();
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java]: If we do the suggested changes below, we should move this to be after the following if else.

File: cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java [hunk 3]
```diff
@@ -187,13 +191,20 @@ public static void disconnectHelixManager() throws Exception {
...
+  }
+
 
-  private Participant(String zkConnectString, String clusterName, String instanceName,
-                     String stateModelType, int port, String postUrl,
-                     boolean useS3Backup, String s3BucketName, boolean runSpectator) throws Exception {
+  public Participant(String zkConnectString, String clusterName, String instanceName,
+                      String stateModelType, int port, String postUrl,
+                      boolean useS3Backup, String s3BucketName, boolean runSpectator)
+      throws Exception {
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java]: why these changes?

Comment (kangnanli) [reply cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java]: I made the constructor to be public for the convenience of my unit testing of verifying registering to multiple modes. 
I removed the unit testing, seem to be unnecessary; thus, change back to "private".

File: cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java [hunk 4]
```diff
@@ -50,6 +53,7 @@
 
 
 public class Participant {
+
```
Comment (newpoo) [inline cluster_management/src/main/java/com/pinterest/rocksplicator/Participant.java]: nit: remove this line.
